### PR TITLE
chore: update webapp to Remix 2.0.1

### LINKED ISSUE #198: [TRI-703] Chore: update webapp to Remix 2.0.1
We're already using most of Remix's v2 features, as future flags in our [remix.config.js](https://github.com/triggerdotdev/trigger.dev/blob/e350659e244e6976747f726984c83a780ba94882/apps/webapp/remix.config.js#L3).

## How to

1. Read Remix's recommendations in the [v2 release](https://github.com/remix-run/remix/releases/tag/remix%402.0.0)
2. They recommend upgrading to 1.19.3 first: `npx upgrade-remix 1.19.3`.
3. Upgrading to `v2_meta` and `v2_headers`. These should be trivial as I don't think we use the headers function anywhere, and every route on the site requires login so only the root.tsx meta function is being used right now.
4. Upgrade Remix to [v2.0.1](https://github.com/remix-run/remix/releases/tag/remix%402.0.1).
5. Run `pnpm run typecheck --filter webapp` to find and fix TypeScript errors.
6. Run the webapp and check it functions as expected

Please follow the [contribution guide](https://github.com/triggerdotdev/trigger.dev/blob/main/CONTRIBUTING.md) development section to setup and run the web app.

<sub>From [SyncLinear.com](https://synclinear.com) | [TRI-703](https://linear.app/triggerdotdev/issue/TRI-703/chore-update-webapp-to-remix-201)</sub>

### PR DESCRIPTION
**Pull request by Dhoni77**
Closes #198

## ✅ Checklist

- [x] I have followed every step in the [contributing guide](https://github.com/triggerdotdev/trigger.dev/blob/main/CONTRIBUTING.md)
- [x] The PR title follows the convention.
- [x] I ran and tested the code works

---

## Changelog

Followed the steps to upgrade as per the remix docs:

- Updated Types and removed deprecated types.
- Modified deprecated useTransition hook to useNavigation hook.
- Updated the Meta function to return array instead of object.
- Remove Future Flags.

### CODE REVIEW

**Comment by Dhoni77** [comment]
cc @ericallam

File: apps/webapp/app/utils.ts [hunk 1]
```diff
@@ -35,8 +35,8 @@ export function safeRedirect(
 export function useMatchesData(
   id: string | string[],
   debug: boolean = false
-): RouteMatch | undefined {
-  const matchingRoutes = useMatches();
+) {
+  const matchingRoutes = useMatches() as UIMatch<unknown, Handle>[];
```
Comment (ericallam) [inline apps/webapp/app/utils.ts:39]: Why is this being cast?

File: apps/webapp/app/components/navigation/Breadcrumb.tsx [hunk 2]
```diff
@@ -1,16 +1,20 @@
...
 
-export type BreadcrumbItem = (match: RouteMatch, allMatches: RouteMatch[]) => ReactNode;
+export type BreadcrumbItem = (
+  match: UIMatch<unknown, Handle>,
+  allMatches: UIMatch<unknown, Handle>[]
+) => ReactNode;
 
 export function Breadcrumb() {
-  const matches = useMatches();
+  const matches: UIMatch<unknown, Handle>[] = useMatches() as UIMatch<unknown, Handle>[];
```
Comment (ericallam) [inline apps/webapp/app/components/navigation/Breadcrumb.tsx]: Why is this being cast? We should avoid doing this kind of thing unless it's absolutely necessary

File: apps/webapp/app/components/run/TaskDetail.tsx [hunk 3]
```diff
@@ -32,7 +32,7 @@ import { ClientOnly } from "remix-utils";
 import { Spinner } from "../primitives/Spinner";
 import type { DetailedTask } from "~/routes/_app.orgs.$organizationSlug.projects.$projectParam.jobs.$jobParam.runs.$runParam.tasks.$taskParam/route";
 
-export function TaskDetail({ task }: { task: DetailedTask }) {
+export function TaskDetail({ task }: { task: DetailedTask & { params: {} } }) {
```
Comment (ericallam) [inline apps/webapp/app/components/run/TaskDetail.tsx:35]: I don't understand why this change is needed

File: apps/webapp/app/components/navigation/ProjectsMenu.tsx [hunk 4]
```diff
@@ -12,8 +12,10 @@ import {
   PopoverMenuItem,
   PopoverSectionHeader,
 } from "../primitives/Popover";
+import { AppData } from "~/utils/appData";
+import { Handle } from "~/utils/handle";
 
-export function ProjectsMenu({ matches }: { matches: RouteMatch[] }) {
+export function ProjectsMenu<T = AppData>({ matches }: { matches: UIMatch<T, Handle>[] }) {
```
Comment (ericallam) [inline apps/webapp/app/components/navigation/ProjectsMenu.tsx]: This is weird, a React component that has type params?

File: apps/webapp/app/hooks/useEnvironments.ts [hunk 5]
```diff
@@ -1,23 +1,25 @@
-import { RouteMatch } from "@remix-run/react";
+import { UIMatch } from "@remix-run/react";
 import { MatchedProject, useOptionalProject } from "./useProject";
+import { AppData } from "~/utils/appData";
+import { Handle } from "~/utils/handle";
 
 export type ProjectJobEnvironment = MatchedProject["environments"][number];
 
-export function useEnvironments(matches?: RouteMatch[]) {
+export function useEnvironments<T = AppData>(matches?: UIMatch<T, Handle>[]) {
```
Comment (ericallam) [inline apps/webapp/app/hooks/useEnvironments.ts]: This seems like it's losing type information here, can you explain what's going on with this type parameter?

**Review by ericallam — CHANGES_REQUESTED** [review]
Would be great if you could address the comments left above and also resolve the conflicts with the main branch so this PR can be easily squashed and merged. Additionally can you make sure that running typecheck and tests work and ideally a screencast added to the PR of the webapp working?

File: apps/webapp/app/utils.ts (hunk 1 above)
Comment (Dhoni77) [reply apps/webapp/app/utils.ts:39]: Hi @ericallam  with the remix update the RouteMatch Type has been removed so we have to use the UIMatch<unknown, unkown> type as a replacement for that.
Since useMatches hook returns UIMatch<unknown, unkown>[] as return type, we won't get type completions and the below paths.reduce function is giving type errors

File: apps/webapp/app/components/run/TaskDetail.tsx (hunk 3 above)
Comment (Dhoni77) [reply apps/webapp/app/components/run/TaskDetail.tsx:35]: It was giving a type error like params property does not exist so had to manually add it.

File: apps/webapp/app/hooks/useEnvironments.ts (hunk 5 above)
Comment (Dhoni77) [reply apps/webapp/app/hooks/useEnvironments.ts]: Since the UIMatch type expects 2 type parameters I had to pass something.
The 2nd parameter is required because we get type completions for the matches from that.
So I'll use (matches?: UIMatch<unknown, Handle>[]) in place of that type param.
Is that ok ?

Comment (ericallam) [reply apps/webapp/app/hooks/useEnvironments.ts]: What's the return type of `useEnvironments` above? And can you run `pnpm run typecheck --filter webapp` and make sure there are no type errors.

Comment (Dhoni77) [reply apps/webapp/app/hooks/useEnvironments.ts]: Hi @ericallam 

typecheck

The type check was successful.

**Comment by Dhoni77** [comment]
cc @ericallam

**Comment by ericallam** [comment]
Pulled this down and looks like there is an issue with the pnpm-lock.yaml file as I'm getting this error when running `pnpm install`:

```sh
Ignoring broken lockfile at /Users/eric/code/triggerdotdev/trigger.dev: Lockfile /Users/eric/code/triggerdotdev/trigger.dev/pnpm-lock.yaml not compatible with current pnpm
```

Can you make sure you are using `pnpm@7.18.1`

**Comment by Dhoni77** [comment]
Hi @ericallam can you try once.
I deleted my old pnpm-lock and reinstalled the dependencies again.

**Comment by ericallam** [comment]
Also ran into this same issue in another PR and left a comment just now about how to recover from this: 

https://github.com/triggerdotdev/trigger.dev/pull/581#issuecomment-1753869773

**Comment by Dhoni77** [comment]
> Also ran into this same issue in another PR and left a comment just now about how to recover from this:
> 
> [#581 (comment)](https://github.com/triggerdotdev/trigger.dev/pull/581#issuecomment-1753869773)

Followed the steps and resolved the issue👍

**Review by ericallam — CHANGES_REQUESTED** [review]
I'm getting the following error in the console when running `pnpm run dev --filter webapp`:

```
webapp:dev: > webapp@1.0.0 dev /Users/eric/code/triggerdotdev/trigger.dev/apps/webapp
webapp:dev: > cross-env PORT=3030 remix dev
webapp:dev: 
webapp:dev: 
webapp:dev:  💿  remix dev
webapp:dev: 
webapp:dev: [info] building...
webapp:dev: ✘ [ERROR] No matching export in "../../node_modules/.pnpm/@remix-run+react@2.0.1_o4scbtliisanygemawej7x2d6i/node_modules/@remix-run/react/dist/esm/index.js" for import "useTransition"
webapp:dev: 
webapp:dev:     ../../node_modules/.pnpm/remix-utils@6.6.0_3g4awgkzsdylkd4ankfzozrqee/node_modules/remix-utils/browser/react/use-global-pending-state.js:1:9:
webapp:dev:       1 │ import { useTransition, useFetchers } from "@remix-run/react";
webapp:dev:         ╵          ~~~~~~~~~~~~~
webapp:dev: 
webapp:dev: 
```

**Comment by ericallam** [comment]
Going to close this now as this PR is quite out of date and the requested change from 2 weeks ago hasn't been addressed. I appreciate the hard work on this PR though and appreciate that it wasn't the easiest one to get over the line as it touched on many different parts of the app.
